File: object_Movie/src/main/java/objectstudy/policy/refund/DefaultRefundPolicy.java
package objectstudy.policy.refund;

import objectstudy.Money;
import objectstudy.Refund;
import objectstudy.Reservation;
import objectstudy.condition.refund.RefundCondition;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

abstract class DefaultRefundPolicy implements RefundPolicy {

    private List<RefundCondition> conditions = new ArrayList<>();

    public DefaultRefundPolicy(RefundCondition... conditions) {

        this.conditions = Arrays.asList(conditions);
    }

    @Override
    public Money calculateRefundAmount(Reservation reservation) {
        for(RefundCondition each : conditions) {
            if (each.isSatisfiedBy(reservation)) {
                return getDiscountAmount(reservation);
            }
        }

        return Money.ZERO;
    }

    abstract protected Money getDiscountAmount(Reservation reservation);
}

File: object/object/src/java_to_kotlin/policy/refund/DefaultRefundPolicy.kt
package java_to_kotlin.policy.refund

import java_to_kotlin.Money
import java_to_kotlin.Refund
import java_to_kotlin.condition.refund.RefundCondition
import java.util.*

abstract class DefaultRefundPolicy(vararg conditions: RefundCondition?) : RefundPolicy {
    private var conditions: MutableList<RefundCondition> = ArrayList()

    init {
        this.conditions = Arrays.asList(*conditions)
    }

    override fun calculateRefundAmount(refund: Refund): Money? {
        for (each in conditions) {
            if (each.isSatisfiedBy(refund)) {
                return getDiscountAmount(refund)
            }
        }
        return Money.Companion.ZERO
    }

    protected abstract fun getDiscountAmount(refund: Refund): Money?
}
